pkg/client/client_test.go

```diff
@@ -1787,34 +1800,276 @@ func TestCreateOrUpdatePrometheus(t *testing.T) {
 				},
 			}
 
+			s := runtime.NewScheme()
+			_ = monv1.AddToScheme(s)
 			c := Client{
-				mclient: monfake.NewSimpleClientset(prometheus.DeepCopy()),
+				dclient:       dynamicFake.NewSimpleDynamicClient(s, prometheus.DeepCopy()),
+				eventRecorder: events.NewInMemoryRecorder("test-create-or-update-prometheus"),
 			}
-			if _, err := c.mclient.MonitoringV1().Prometheuses(ns).Get(ctx, prometheus.GetName(), metav1.GetOptions{}); err != nil {
+			if _, err := c.dclient.Resource(monv1.SchemeGroupVersion.WithResource("prometheuses")).Namespace(ns).Get(ctx, prometheus.GetName(), metav1.GetOptions{}); err != nil {
 				t.Fatal(err)
 			}
 
 			prometheus.SetLabels(tc.updatedLabels)
 			prometheus.SetAnnotations(tc.updatedAnnotations)
-			if err := c.CreateOrUpdatePrometheus(ctx, prometheus); err != nil {
+			if _, err := c.CreateOrUpdatePrometheus(ctx, prometheus); err != nil {
 				t.Fatal(err)
 			}
 
-			after, err := c.mclient.MonitoringV1().Prometheuses(ns).Get(ctx, prometheus.GetName(), metav1.GetOptions{})
+			after, err := c.dclient.Resource(monv1.SchemeGroupVersion.WithResource("prometheuses")).Namespace(ns).Get(ctx, prometheus.GetName(), metav1.GetOptions{})
 			if err != nil {
 				t.Fatal(err)
 			}
 
-			if !reflect.DeepEqual(tc.expectedAnnotations, after.Annotations) {
-				t.Errorf("expected annotations %q, got %q", tc.expectedAnnotations, after.Annotations)
+			if !reflect.DeepEqual(tc.expectedAnnotations, after.GetAnnotations()) {
+				t.Errorf("expected annotations %q, got %q", tc.expectedAnnotations, after.GetAnnotations())
 			}
-			if !reflect.DeepEqual(tc.expectedLabels, after.Labels) {
-				t.Errorf("expected labels %q, got %q", tc.expectedLabels, after.Labels)
+			if !reflect.DeepEqual(tc.expectedLabels, after.GetLabels()) {
+				t.Errorf("expected labels %q, got %q", tc.expectedLabels, after.GetLabels())
+			}
+		})
+	}
+
+	initialPrometheus := &monv1.Prometheus{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Prometheus",
+			APIVersion: "monitoring.coreos.com/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-prometheus",
+			Namespace: "test-namespace",
+		},
+	}
+	prometheusGVR := initialPrometheus.GroupVersionKind().GroupVersion().WithResource("prometheuses")
+
+	c := &Client{}
+	c.dclient = dynamicFake.NewSimpleDynamicClient(runtime.NewScheme(), initialPrometheus.DeepCopy())
+	c.eventRecorder = events.NewInMemoryRecorder("create-or-update-prometheus-test")
+	c.resourceCache = resourceapply.NewResourceCache()
```

Comment:
Note that the code seems to be using the "comparison approach" (I don't see any cache initialized there), but here we're testing the "cache approach", am I missing sth?

Reply:
Both of these are in use (PTAL my comment https://github.com/openshift/cluster-monitoring-operator/pull/2395/files#r1696698968). Here, we initialize the cache, and down below we use the defaulting approach to reject default `spec` updates (`shouldUpdate` is `false` for [this case](https://github.com/openshift/cluster-monitoring-operator/pull/2395/files#diff-2dad64e28ce5e2eddf0adf303052d38e3b5b4fcba6b32fc4ca26d5e4d6b8fa4aR1861)).